### DISCUSSION TITLE
RNNs/LSTMs and FINN

### DISCUSSION THREAD

**Discussion by fannymonori** [Xilinx/finn#532] [Q&A] [answered]
Hi everyone,

I really appreciate the work you do with the FINN framework. I'm curious, does FINN support recurrent neural networks or LSTMs? If not, would the current FINN infrastructure theoretically allow it to be extended with recurrent layers?

Many thanks,
Fanny

**Comment by heborras** [comment, marked as answer]
Hi @fannymonori ,

unfortunately FINN does currently not support RNNs or LSTMs.

However, in theory FINN can indeed be extended to support recurrent structures. And one would probably need to implement a few new HLS operators (see: https://github.com/Xilinx/finn/blob/main/notebooks/advanced/2_custom_op.ipynb ) and transformation passes (see: https://github.com/Xilinx/finn/blob/main/notebooks/advanced/1_custom_transformation_pass.ipynb ). Along with the required development for the actual HLS functions this undertaking would be quite involved, but possible.

Cheers,
Hendrik

**Reply by fannymonori** [reply]
Thank you Hendrik! Your answer was helpful nonetheless.
